### PR TITLE
Make it so droids be used as supply beacon

### PR DESCRIPTION
**Pull request by lbnesquik**
## About The Pull Request
This PR allows the person controlling a droid to use it as an antenna. 
It is restricted to the humanoid droids, not the little tank guys.
It works for someone with a remote and the AI

## Why It's Good For The Game
I feel like it's a cool idea.
Having the humanoid droids be able to do more stuff than just shoot, compared to the 'dumber' gun vehicles. 

I was thinking of having it be just for the scout droid, but i'd rather have some feedback on it first. 
Having more abilities tied to droids would make them less dangerous for xenos, as risking them is a bigger waste. 

## Changelog
:cl:
add: Made it so droids can be used as supply beacons. 
/:cl:

### CODE REVIEW

**Comment by MereWalker** [comment]
Go ahead and give AI a droid with hands. So it can help out on the ship. Please I need hands.

File: code/modules/vehicles/unmanned/unmanned_droid.dm [hunk 1]
```diff
@@ -83,3 +95,35 @@
...
+/obj/vehicle/unmanned/droid/proc/clean_beacon_datum()
+	SIGNAL_HANDLER
+	beacon_datum = null
+
+/datum/action/antenna
+	name = "Use Antenna"
+	action_icon_state = "rally_minions"
+
+/datum/action/antenna/action_activate()
+	SEND_GLOBAL_SIGNAL(COMSIG_GLOB_UNMANNED_COORDINATES)
```
[the remark this answers is not in the document]
Comment (BraveMole) [reply code/modules/vehicles/unmanned/unmanned_droid.dm:129]: A signal on owner should do the trick at first glance

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/vehicles/unmanned/unmanned_droid.dm [hunk 2]
```diff
@@ -83,3 +96,33 @@
 	playsound(src, 'sound/effects/pred_cloakoff.ogg', 60, TRUE)
 	alpha = initial(alpha)
 	TIMER_COOLDOWN_START(src, COOLDOWN_DRONE_CLOAK, 12 SECONDS)
+
+///Proc used for the supply link feature, activate to appear as an antenna
+/obj/vehicle/unmanned/droid/proc/activate_antenna(datum/source, mob/user)
+	SIGNAL_HANDLER
+
+	for(var/mob/temp_user AS in client_mobs_in_contents)
+		user = temp_user
```
Comment (BraveMole) [inline code/modules/vehicles/unmanned/unmanned_droid.dm:105]: what is going on here

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/__DEFINES/dcs/signals.dm [hunk 3]
```diff
@@ -253,6 +253,8 @@
 
 #define COMSIG_UNMANNED_TURRET_UPDATED "unmanned_turret_update" /// from /obj/vehicle/unmanned/attackby: (newtype)
 #define COMSIG_UNMANNED_ABILITY_UPDATED "unmanned_ability_update"
+#define COMSIG_GLOB_UNMANNED_COORDINATES "unmanned_coordinates"
+#define COMSIG_UNMANNED_COORDINATES "unmanned_coordinates"
```
Comment (TiviPlus) [inline code/__DEFINES/dcs/signals.dm:257]: dupe signals

File: code/modules/vehicles/unmanned/unmanned_droid.dm [hunk 4]
```diff
@@ -9,6 +9,9 @@
 	gunnoise = 'sound/weapons/guns/fire/laser.ogg'
 	spawn_equipped_type = /obj/item/uav_turret/droid
 	unmanned_flags = HAS_LIGHTS|OVERLAY_TURRET
+	/// Reference to the datum used by the supply drop console
+	var/datum/supply_beacon/beacon_datum
+	var/datum/action/antenna/antenna = new
```
Comment (TiviPlus) [inline code/modules/vehicles/unmanned/unmanned_droid.dm:14]: missing autodoc and why isnt this new in Initialize

File: code/modules/vehicles/unmanned/unmanned_droid.dm [hunk 5]
```diff
@@ -83,3 +96,33 @@
...
+
+	if(beacon_datum)
+		UnregisterSignal(beacon_datum, COMSIG_PARENT_QDELETING)
+		QDEL_NULL(beacon_datum)
+		to_chat(user, (span_warning("The [src] beeps and states, \"Your last position is no longer accessible by the supply console")))
+		return
+	if(!is_ground_level(src.z))
+		to_chat(user, span_warning("You have to be on the planet to use this or it won't transmit."))
+		return FALSE
+	beacon_datum = new /datum/supply_beacon(user.name, src.loc, user.faction, 4 MINUTES)
```
Comment (TiviPlus) [inline code/modules/vehicles/unmanned/unmanned_droid.dm:115]: src.var, remove please

File: code/modules/vehicles/unmanned/unmanned_droid.dm [hunk 6]
```diff
@@ -83,3 +95,35 @@
...
+	RegisterSignal(beacon_datum, COMSIG_PARENT_QDELETING, .proc/clean_beacon_datum)
+	user.show_message(span_notice("The [src] beeps and states, \"Your current coordinates were registered by the supply console. LONGITUDE [location.x]. LATITUDE [location.y]. Area ID: [get_area(src)]\""))
+
+/obj/vehicle/unmanned/droid/proc/clean_beacon_datum()
```
[the remark this answers is not in the document]
Comment (TiviPlus) [reply code/modules/vehicles/unmanned/unmanned_droid.dm:119]: dont resolve without fixing issues please

File: code/modules/vehicles/unmanned/unmanned_droid.dm [hunk 7]
```diff
@@ -83,3 +96,33 @@
...
+	SIGNAL_HANDLER
+	beacon_datum = null
+
+/datum/action/antenna
+	name = "Use Antenna"
+	action_icon_state = "rally_minions"
+
+/datum/action/antenna/action_activate()
+	SEND_SIGNAL(owner, COMSIG_UNMANNED_COORDINATES)
```
Comment (TiviPlus) [inline code/modules/vehicles/unmanned/unmanned_droid.dm:128]: I dont know what to think about this signal

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/vehicles/unmanned/unmanned_droid.dm (hunk 4 above)
Comment (BraveMole) [reply code/modules/vehicles/unmanned/unmanned_droid.dm:13]: It is extremly extremly unlikely (impossible) that this prevents compilation.

Declare the action here, but say it is equal to new in init

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/vehicles/unmanned/unmanned_droid.dm (hunk 7 above)
Comment (BraveMole) [reply code/modules/vehicles/unmanned/unmanned_droid.dm:128]: Looks fine to me, you cannot really get the droid from the owner

File: code/modules/vehicles/unmanned/unmanned_droid.dm (hunk 6 above)
[the remark this answers is not in the document]
Comment (BraveMole) [reply code/modules/vehicles/unmanned/unmanned_droid.dm:123]: it's ///, not //

File: code/modules/vehicles/unmanned/unmanned_droid.dm [hunk 8]
```diff
@@ -25,11 +31,21 @@
...
 	else
 		playsound(src, 'sound/machines/drone/droneoff.ogg', 70)
 		STOP_PROCESSING(SSslowprocess, src)
 		user.clear_fullscreen("machine", 5)
+		antenna.remove_action(user)
+		UnregisterSignal(user, COMSIG_UNMANNED_COORDINATES)
 
+/obj/vehicle/unmanned/droid/Destroy(datum/source, mob/user)
+	. = ..()
+	GLOB.unmanned_vehicles -= src
```
Comment (BraveMole) [inline code/modules/vehicles/unmanned/unmanned_droid.dm:45]: Isnt that already being done in vehicle unmanned ?

File: code/modules/vehicles/unmanned/unmanned_droid.dm [hunk 9]
```diff
@@ -83,3 +99,33 @@
...
+		to_chat(source, (span_warning("The [src] beeps and states, \"Your last position is no longer accessible by the supply console")))
+		return
+	if(!is_ground_level(z))
+		to_chat(source, span_warning("You have to be on the planet to use this or it won't transmit."))
+		return FALSE
+	beacon_datum = new /datum/supply_beacon(user.name, src.loc, user.faction, 4 MINUTES)
+	RegisterSignal(beacon_datum, COMSIG_PARENT_QDELETING, .proc/clean_beacon_datum)
+	user.show_message(span_notice("The [src] beeps and states, \"Your current coordinates were registered by the supply console. LONGITUDE [loc.x]. LATITUDE [loc.y]. Area ID: [get_area(src)]\""))
+
+//removes the beacon when we delete the droid
```
Comment (BraveMole) [inline code/modules/vehicles/unmanned/unmanned_droid.dm:121]:
```suggestion
///removes the beacon when we delete the droid
```
Like this

**Review by BraveMole — APPROVED** [review]

File: code/modules/vehicles/unmanned/unmanned_droid.dm [hunk 10]
```diff
@@ -9,6 +9,12 @@
 	gunnoise = 'sound/weapons/guns/fire/laser.ogg'
 	spawn_equipped_type = /obj/item/uav_turret/droid
 	unmanned_flags = HAS_LIGHTS|OVERLAY_TURRET
+	var/datum/supply_beacon/beacon_datum
+	var/datum/action/antenna/antenna
```
Comment (BraveMole) [inline code/modules/vehicles/unmanned/unmanned_droid.dm:13]: autodoc

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/vehicles/unmanned/unmanned_droid.dm (hunk 10 above)
Comment (BraveMole) [reply code/modules/vehicles/unmanned/unmanned_droid.dm:14]: U need to autodoc every var

File: code/modules/vehicles/unmanned/unmanned_droid.dm [hunk 11]
```diff
@@ -25,11 +32,20 @@
...
+		antenna.give_action(user)
+		RegisterSignal(user, COMSIG_UNMANNED_COORDINATES, .proc/activate_antenna)
 	else
 		playsound(src, 'sound/machines/drone/droneoff.ogg', 70)
 		STOP_PROCESSING(SSslowprocess, src)
 		user.clear_fullscreen("machine", 5)
+		antenna.remove_action(user)
+		UnregisterSignal(user, COMSIG_UNMANNED_COORDINATES)
 
+/obj/vehicle/unmanned/droid/Destroy(datum/source, mob/user)
```
Comment (BraveMole) [inline code/modules/vehicles/unmanned/unmanned_droid.dm:44]: In a destroy, you have to finish by return ..() rather than starting by ..()

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/vehicles/unmanned/unmanned_droid.dm [hunk 12]
```diff
@@ -83,3 +100,33 @@
...
+	user.show_message(span_notice("The [src] beeps and states, \"Your current coordinates were registered by the supply console. LONGITUDE [loc.x]. LATITUDE [loc.y]. Area ID: [get_area(src)]\""))
+
+///removes the beacon when we delete the droid
+/obj/vehicle/unmanned/droid/proc/clean_beacon_datum()
+	SIGNAL_HANDLER
+	beacon_datum = null
+
+/datum/action/antenna
+	name = "Use Antenna"
+	action_icon_state = "rally_minions"
```
Comment (BraveMole) [inline code/modules/vehicles/unmanned/unmanned_droid.dm:129]: You probably need a custom icon, doesnt this one have xenos on it ?

**Review by BraveMole — APPROVED** [review]

**Pull request merged**